internal/db/schema/migrations/oss/postgres/101/01_app_token_base.up.sql

```diff
@@ -0,0 +1,246 @@
+  begin
+    insert into app_token
+      (public_id, scope_id, expiration_time, time_to_stale_seconds)
+    values
+      (new.public_id, new.scope_id, new.expiration_time, new.time_to_stale_seconds);
+    return new;
+  end;
+  $$ language plpgsql;
+  comment on function insert_app_token_subtype() is
+    'insert_app_token_subtype is used to automatically insert a row into the app_token table '
+    'whenever a row is inserted into the subtype table';
+
+-- Add trigger to update the new update_time column on every app_token subtype update.
+  create function update_app_token_table_update_time() returns trigger
+  as $$
+  begin
+    update app_token
+       set update_time = new.update_time
+     where public_id = new.public_id;
+    return new;
+  end;
+  $$ language plpgsql;
+  comment on function update_app_token_table_update_time() is
+    'update_app_token_table_update_time is used to automatically update the update_time '
+      'of the base table whenever one of the subtype app_token tables are updated';
+
+-- Add trigger to update the new approximate_last_access_time column on every app_token subtype update.
+  create function update_app_token_table_approximate_last_access_time() returns trigger
+  as $$
+  begin
+    -- Only update if approximate_last_access_time has actually changed
+    if old.approximate_last_access_time is distinct from new.approximate_last_access_time then
+      update app_token
+         set approximate_last_access_time = new.approximate_last_access_time
+       where public_id = new.public_id;
+    end if;
+    return new;
+  end;
+  $$ language plpgsql;
+  comment on function update_app_token_table_approximate_last_access_time() is
+    'update_app_token_table_approximate_last_access_time is used to automatically update the approximate_last_access_time '
+      'of the base table whenever one of the subtype app_token tables are updated';
+
+  -- Function to validate that created_by_user_id exists in iam_user
+  create or replace function validate_app_token_created_by_user() returns trigger
+  as $$
+  begin
+    perform 1
+      from iam_user
+     where public_id = new.created_by_user_id;
+    if not found then
+      raise exception 'User ID % does not exist in iam_user', new.created_by_user_id;
+    end if;
+    return new;
+  end;
+  $$ language plpgsql;
+  comment on function validate_app_token_created_by_user() is
+    'validate_app_token_created_by_user is used to enforce that created_by_user_id exists in iam_user table';
+
+  -- Function to validate that revoked can only be updated from false to true
+  create or replace function validate_app_token_revocation() returns trigger
+  as $$
+  begin
+    -- For updates, check revoked field changes
+    if old.revoked is distinct from new.revoked then
+      -- Only allow change from false to true
+      if not (old.revoked = false and new.revoked = true) then
+        raise exception 'App token cannot be unrevoked. revoked value. Current: %, Attempted: %',
```

Comment:
It looks like we don't need the "revoked value." part of this exception message

```suggestion
        raise exception 'App token cannot be unrevoked. Current: %, Attempted: %',
```